src/v/utils/chunked_kv_cache.h

```diff
@@ -0,0 +1,195 @@
+      : _cache{config, evict{*this}} {}
+
+    ~chunked_kv_cache() noexcept = default;
+
+    // These contructors need to be deleted to ensure a stable `this` pointer.
+    chunked_kv_cache(chunked_kv_cache&&) = delete;
+    chunked_kv_cache& operator=(chunked_kv_cache&&) noexcept = delete;
+    chunked_kv_cache(const chunked_kv_cache&) = delete;
+    chunked_kv_cache& operator=(const chunked_kv_cache&) noexcept = delete;
+
+    /**
+     * Inserts a value for a given key into the cache.
+     *
+     * Returns true if the value was inserted and false if there was already a
+     * value for the given key in the cache.
+     */
+    bool try_insert(const Key& key, ss::shared_ptr<Value> val);
+
+    /**
+     * Gets the key's corresponding value from the cache.
+     *
+     * Returns std::nullopt if the key doesn't have a value in the cache.
+     */
+    ss::optimized_optional<ss::shared_ptr<Value>> get_value(const Key& key);
+
+    using cache_stat = struct cache_t::stat;
+    /**
+     * Cache statistics.
+     */
+    struct stat : public cache_stat {
+        /// Current size of the cache index.
+        size_t index_size;
+    };
+
+    /**
+     * Returns the current cache statistics.
+     */
+    [[nodiscard]] stat stat() const noexcept;
+
+private:
+    using ghost_hook_t = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::safe_link>>;
+
+    struct cached_value {
+        Key key;
+        ss::shared_ptr<Value> value;
+        s3_fifo::cache_hook hook;
+        ghost_hook_t ghost_hook;
+    };
+
+    using entry_t = std::unique_ptr<cached_value>;
+    using ghost_fifo_t = boost::intrusive::list<
+      cached_value,
+      boost::intrusive::
+        member_hook<cached_value, ghost_hook_t, &cached_value::ghost_hook>>;
+
+    chunked_hash_map<Key, entry_t, Hash, EqualTo> _map;
+    cache_t _cache;
+    ghost_fifo_t _ghost_fifo;
+
+    void gc_ghost_fifo();
+};
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+struct chunked_kv_cache<Key, Value, Hash, EqualTo>::evict {
+    chunked_kv_cache& kv_c;
+
+    bool operator()(cached_value& e) noexcept {
+        e.value = nullptr;
+        kv_c._ghost_fifo.push_back(e);
+        return true;
+    }
+};
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+bool chunked_kv_cache<Key, Value, Hash, EqualTo>::try_insert(
+  const Key& key, ss::shared_ptr<Value> val) {
+    gc_ghost_fifo();
+
+    auto e_it = _map.find(key);
+    if (e_it == _map.end()) {
+        auto [e_it, succ] = _map.try_emplace(
+          key, std::make_unique<cached_value>(key, std::move(val)));
+        if (!succ) {
+            return false;
+        }
+
+        _cache.insert(*e_it->second);
+        return true;
+    }
+
+    auto& entry = *e_it->second;
+    if (entry.hook.evicted()) {
+        entry.value = std::move(val);
+        _ghost_fifo.erase(_ghost_fifo.iterator_to(entry));
+        _cache.insert(entry);
```

Comment:
does insert reset the freq (that's bumped by `touch`) -- can't remember if there was something special that happened when entry was promoted from ghost to normal queues, or if it sort of picked up where it left off. actually yeh, i think that later interpretation rings a bell.

Reply:
Yeah, I have a test for the cache that ensures that the entry is promoted from the ghost queue to the main queue in the case handled by this section of code.